### PR TITLE
Automatic table of contents generation for Markdown

### PR DESCRIPTION
**Pull request by mbarczak**
This PR aims to solve #1436.

- Adds table of contents for any currently displayed Markdown note or Markdown snippet, at current cursor position.
- Consequent generations update existing TOC.
- Generated TOC is case sensitive to handle #2067

**Shortcut** : _SHIFT+CTRL+T_
**Menu** : _File / Generate/Update Markdown TOC_

### CODE REVIEW

**Comment by ZeroX-DG** [comment]
LGTM :tada:

File: browser/main/Detail/MarkdownNoteDetail.js [hunk 1]
```diff
@@ -262,6 +266,11 @@ class MarkdownNoteDetail extends React.Component {
     }
   }
 
+  handleGenerateToc () {
+    markdownToc.generate(this.refs.content.value,
+      (modifiedValue) => { this.refs.content.refs.code.setValue(modifiedValue) })
```
Comment (ZeroX-DG) [inline browser/main/Detail/MarkdownNoteDetail.js]: Can you remove the `{` and `}` to retain the code style

File: browser/main/Detail/SnippetNoteDetail.js [hunk 2]
```diff
@@ -91,6 +94,16 @@ class SnippetNoteDetail extends React.Component {
...
     if (this.saveQueue != null) this.saveNow()
+    ee.off('code:generate-toc', this.generateToc)
+  }
+
+  handleGenerateToc () {
+    const currentMode = this.state.note.snippets[this.state.snippetIndex].mode
+    if (currentMode.includes('Markdown')) {
+      const currentValue = this.refs['code-' + this.state.snippetIndex].value
+      const currentEditor = this.refs['code-' + this.state.snippetIndex].refs.code.editor
+      markdownToc.generate(currentValue, (modifiedValue) => { currentEditor.setValue(modifiedValue) })
```
Comment (ZeroX-DG) [inline browser/main/Detail/SnippetNoteDetail.js]: Can you remove the `{` and `}` to retain the code style. Like this:
```
(modifiedValue) => currentEditor.setValue(modifiedValue)
```

Comment (mbarczak) [reply browser/main/Detail/SnippetNoteDetail.js]: Done 👍

File: browser/main/Detail/MarkdownNoteDetail.js (hunk 1 above)
Comment (mbarczak) [reply browser/main/Detail/MarkdownNoteDetail.js]: 👍

**Comment by ZeroX-DG** [comment]
Hey @mbarczak , just one more small request, it would be great if you can insert the toc at the cursor position. Sometimes, I want to have a document like this:
```
This is some introduction...
<!--TOC-->
# h1
## h2
```
With the current code it only insert at the top of the document.

**Comment by mbarczak** [comment]
@ZeroX-DG , I've introduced change you requested, and indeed, it's more usable now. Also, I've added additional tests with CodeMirror mock.

File: browser/main/Detail/SnippetNoteDetail.js [hunk 3]
```diff
@@ -91,6 +94,15 @@ class SnippetNoteDetail extends React.Component {
 
   componentWillUnmount () {
     if (this.saveQueue != null) this.saveNow()
+    ee.off('code:generate-toc', this.generateToc)
+  }
+
+  handleGenerateToc () {
+    const currentMode = this.state.note.snippets[this.state.snippetIndex].mode
```
Comment (ZeroX-DG) [inline browser/main/Detail/SnippetNoteDetail.js]: Can you use `Object destructuring` to retain the code style?
```
const { note, snippetIndex } = this.state
const currentMode = note.snippets[snippetIndex].mode
```

File: browser/main/Detail/SnippetNoteDetail.js [hunk 4]
```diff
@@ -91,6 +94,15 @@ class SnippetNoteDetail extends React.Component {
...
+    if (currentMode.includes('Markdown')) {
+      const currentEditor = this.refs['code-' + this.state.snippetIndex].refs.code.editor
```
Comment (ZeroX-DG) [inline browser/main/Detail/SnippetNoteDetail.js]: Here, you can use template string:
```
this.refs[`code-${snippetIndex}`].refs.code.editor
```

**Comment by mbarczak** [comment]
@ZeroX-DG , all requests solved ;)

**Comment by mbarczak** [comment]
Hi @Rokt33r , work is done here. Ready to be merged.

**Comment by kazup01** [comment]
Hi @mbarczak, I featured your contribution on release blog 😄 Thank you!
https://medium.com/boostnote/boostnote-v0-11-10-release-on-halloween-9bbe6d488fe8

**Pull request merged**